Flaky test_try_premium_at_start_old_account_can_pull_old_data

## Problem Definition

Test test_try_premium_at_start_old_account_can_pull_old_data seems to be flaky. Failed in OSX Travis.

https://travis-ci.org/github/rotki/rotki/jobs/675324686#L5056

=================================== FAILURES ===================================
________ test_try_premium_at_start_old_account_can_pull_old_data[True] _________
rotkehlchen_instance = <rotkehlchen.rotkehlchen.Rotkehlchen object at 0x10e10e048>
username = 'testuser', db_password = '123'
rotki_premium_credentials = <rotkehlchen.premium.premium.PremiumCredentials object at 0x10dc745f8>
    @pytest.mark.parametrize('start_with_valid_premium', [True])
    def test_try_premium_at_start_old_account_can_pull_old_data(
            rotkehlchen_instance,
            username,
            db_password,
            rotki_premium_credentials,
    ):
        """
        Test that if the remote DB is of an old version its upgraded before replacing our current
    
        For an old account
        """
        setup_starting_environment(
            rotkehlchen_instance=rotkehlchen_instance,
            username=username,
            db_password=db_password,
            premium_credentials=rotki_premium_credentials,
            first_time=False,
            same_hash_with_remote=False,
            newer_remote_db=True,
            db_can_sync_setting=True,
            remote_data=REMOTE_DATA_OLDER_DB,
        )
>       assert_db_got_replaced(rotkehlchen_instance=rotkehlchen_instance, username=username)
rotkehlchen/tests/integration/test_premium.py:247: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
rotkehlchen_instance = <rotkehlchen.rotkehlchen.Rotkehlchen object at 0x10e10e048>
username = 'testuser'
    def assert_db_got_replaced(rotkehlchen_instance: Rotkehlchen, username: str):
        """For environment setup with setup_starting_environment make sure DB is replaced"""
        msg = 'Test default main currency should be different from the restored currency'
        assert DEFAULT_TESTS_MAIN_CURRENCY != A_GBP, msg
        # At this point pulling data from rotkehlchen server should have worked
        # and our database should have been replaced. The new data have different
        # main currency
>       assert rotkehlchen_instance.data.db.get_main_currency() == A_GBP
E       AssertionError

## Task

Investigate and fix if possible

--- a/docs/api.rst
+++ b/docs/api.rst
@@ -146,7 +146,7 @@ Handling user creation, sign-in, log-out and querying
    :resjson object result: For succesful requests, result contains the currently connected exchanges, and the user's settings. For details on the user settings refer to the `Getting or modifying settings`_ section.
    :statuscode 200: Adding the new user was succesful
    :statuscode 400: Provided JSON is in some way malformed
-   :statuscode 409: User already exists. Another user is already logged in. Given Premium API credentials are invalid.
+   :statuscode 409: User already exists. Another user is already logged in. Given Premium API credentials are invalid. Permission error while trying to access the directory where Rotki saves data.
    :statuscode 500: Internal Rotki error
 
 .. http:patch:: /api/(version)/users/(username)
@@ -207,7 +207,7 @@ Handling user creation, sign-in, log-out and querying
    :statuscode 300: Possibility of syncing exists and the login was sent with sync_approval set to ``"unknown"``. Consumer of api must resend with ``"yes"`` or ``"no"``.
    :statuscode 400: Provided JSON is in some way malformed
    :statuscode 401: Provided password is wrong for the user or some other authentication error.
-   :statuscode 409: Another user is already logged in. User does not exist. There was a fatal error during the upgrade of the DB.
+   :statuscode 409: Another user is already logged in. User does not exist. There was a fatal error during the upgrade of the DB. Permission error while trying to access the directory where Rotki saves data.
    :statuscode 500: Internal Rotki error
 
 .. http:patch:: /api/(version)/users/(username)

--- a/rotkehlchen/__main__.py
+++ b/rotkehlchen/__main__.py
@@ -1,7 +1,8 @@
 from gevent import monkey  # isort:skip # noqa
 monkey.patch_all()  # isort:skip # noqa
-
 import logging
+
+from rotkehlchen.errors import SystemPermissionError
 
 logger = logging.getLogger(__name__)
 
@@ -12,6 +13,9 @@ def main() -> None:
     from rotkehlchen.server import RotkehlchenServer
     try:
         rotkehlchen_server = RotkehlchenServer()
+    except SystemPermissionError as e:
+        print(f'ERROR at initialization: {str(e)}')
+        sys.exit(1)
     except SystemExit as e:
         # Mypy here thinks e.code is always None which is not correct according to the docs:
         # https://docs.python.org/3/library/exceptions.html#SystemExit

--- a/rotkehlchen/api/rest.py
+++ b/rotkehlchen/api/rest.py
@@ -33,6 +33,7 @@ from rotkehlchen.errors import (
     PremiumAuthenticationError,
     RemoteError,
     RotkehlchenPermissionError,
+    SystemPermissionError,
     TagConstraintError,
 )
 from rotkehlchen.exchanges.data_structures import Trade
@@ -807,7 +808,7 @@ class RestAPI():
             )
         # not catching RotkehlchenPermissionError here as for new account with premium
         # syncing there is no way that permission needs to be asked by the user
-        except (AuthenticationError, PremiumAuthenticationError) as e:
+        except (AuthenticationError, PremiumAuthenticationError, SystemPermissionError) as e:
             self.rotkehlchen.reset_after_failed_account_creation_or_login()
             result_dict['message'] = str(e)
             return api_response(result_dict, status_code=HTTPStatus.CONFLICT)
@@ -853,7 +854,7 @@ class RestAPI():
             self.rotkehlchen.reset_after_failed_account_creation_or_login()
             result_dict['message'] = str(e)
             return api_response(result_dict, status_code=HTTPStatus.MULTIPLE_CHOICES)
-        except DBUpgradeError as e:
+        except (DBUpgradeError, SystemPermissionError) as e:
             self.rotkehlchen.reset_after_failed_account_creation_or_login()
             result_dict['message'] = str(e)
             return api_response(result_dict, status_code=HTTPStatus.CONFLICT)

--- a/rotkehlchen/data_handler.py
+++ b/rotkehlchen/data_handler.py
@@ -14,7 +14,7 @@ from rotkehlchen.constants.misc import ZERO
 from rotkehlchen.crypto import decrypt, encrypt
 from rotkehlchen.datatyping import BalancesData
 from rotkehlchen.db.dbhandler import DBHandler
-from rotkehlchen.errors import AuthenticationError
+from rotkehlchen.errors import AuthenticationError, SystemPermissionError
 from rotkehlchen.logging import RotkehlchenLogsAdapter
 from rotkehlchen.typing import AssetAmount, B64EncodedBytes, B64EncodedString, FilePath, Timestamp
 from rotkehlchen.user_messages import MessagesAggregator
@@ -48,12 +48,26 @@ class DataHandler():
             password: str,
             create_new: bool,
     ) -> FilePath:
+        """Unlocks a user, either logging them in or creating a new user
+
+        May raise:
+        - SystemPermissionError if there are permission errors when accessing the DB
+        or a directory in the user's filesystem
+        - AuthenticationError if the given user does not exist, or if
+        sqlcipher version problems are detected
+        - DBUpgradeError if the rotki DB version is newer than the software or
+        there is a DB upgrade and there is an error.
+        """
         user_data_dir = FilePath(os.path.join(self.data_directory, username))
         if create_new:
             if os.path.exists(user_data_dir):
                 raise AuthenticationError('User {} already exists'.format(username))
             else:
-                os.mkdir(user_data_dir)
+                try:
+                    os.mkdir(user_data_dir)
+                except PermissionError as e:
+                    raise SystemPermissionError(f'Failed to create directory for user: {str(e)}')
+
         else:
             if not os.path.exists(user_data_dir):
                 raise AuthenticationError('User {} does not exist'.format(username))
@@ -71,9 +85,10 @@ class DataHandler():
                     ),
                 )
 
-                raise AuthenticationError(
+                raise SystemPermissionError(
                     'User {} exists but DB is missing. Somehow must have been manually '
-                    'deleted or is corrupt. Please recreate the user account. '
+                    'deleted or is corrupt or access permissions do not allow reading. '
+                    'Please recreate the user account. '
                     'A backup of the user directory was created.'.format(username))
 
         self.db: DBHandler = DBHandler(user_data_dir, password, self.msg_aggregator)
@@ -145,8 +160,15 @@ class DataHandler():
         users = {}
         data_dir = Path(self.data_directory)
         for x in data_dir.iterdir():
-            if x.is_dir() and (x / 'rotkehlchen.db').exists():
-                users[x.stem] = 'loggedin' if x.stem == self.username else 'loggedout'
+            try:
+                if x.is_dir() and (x / 'rotkehlchen.db').exists():
+                    users[x.stem] = 'loggedin' if x.stem == self.username else 'loggedout'
+            except PermissionError:
+                # __import__("pdb").set_trace()
+
+                # ignore directories that can't be accessed
+                continue
+
         return users
 
     def set_fiat_balances(
@@ -198,6 +220,7 @@ class DataHandler():
         - UnableToDecryptRemoteData due to decrypt()
         - DBUpgradeError if the rotki DB version is newer than the software or
         there is a DB upgrade and there is an error.
+        - SystemPermissionError if the DB file permissions are not correct
         """
         log.info('Decompress and decrypt DB')
 

--- a/rotkehlchen/data_handler.py
+++ b/rotkehlchen/data_handler.py
@@ -164,8 +164,6 @@ class DataHandler():
                 if x.is_dir() and (x / 'rotkehlchen.db').exists():
                     users[x.stem] = 'loggedin' if x.stem == self.username else 'loggedout'
             except PermissionError:
-                # __import__("pdb").set_trace()
-
                 # ignore directories that can't be accessed
                 continue
 

--- a/rotkehlchen/db/dbhandler.py
+++ b/rotkehlchen/db/dbhandler.py
@@ -41,6 +41,7 @@ from rotkehlchen.errors import (
     DeserializationError,
     IncorrectApiKeyFormat,
     InputError,
+    SystemPermissionError,
     TagConstraintError,
     UnknownAsset,
     UnsupportedAsset,
@@ -157,6 +158,7 @@ class DBHandler:
         - DBUpgradeError if the rotki DB version is newer than the software or
         there is a DB upgrade and there is an error.
         - AuthenticationError if SQLCipher version problems are detected
+        - SystemPermissionError if the DB file's permissions are not correct
         """
         self.msg_aggregator = msg_aggregator
         self.user_data_dir = user_data_dir
@@ -188,7 +190,13 @@ class DBHandler:
 
     def __del__(self) -> None:
         self.disconnect()
-        dbinfo = {'sqlcipher_version': self.sqlcipher_version, 'md5_hash': self.get_md5hash()}
+        try:
+            dbinfo = {'sqlcipher_version': self.sqlcipher_version, 'md5_hash': self.get_md5hash()}
+        except SystemPermissionError as e:
+            # If there is problems opening the DB at destruction just log and exit
+            log.error(f'At DB teardown could not open the DB: {str(e)}')
+            return
+
         with open(os.path.join(self.user_data_dir, DBINFO_FILENAME), 'w') as f:
             f.write(rlk_jsondumps(dbinfo))
 
@@ -224,18 +232,31 @@ class DBHandler:
         DBUpgradeManager(self).run_upgrades()
 
     def get_md5hash(self) -> str:
+        """Get the md5hash of the DB
+
+        May raise:
+        - SystemPermissionError if there are permission errors when accessing the DB
+        """
         no_active_connection = not hasattr(self, 'conn') or not self.conn
         assert no_active_connection, 'md5hash should be taken only with a closed DB'
         filename = os.path.join(self.user_data_dir, 'rotkehlchen.db')
         md5_hash = hashlib.md5()
-        with open(filename, 'rb') as f:
-            # Read and update hash in chunks of 4K
-            for byte_block in iter(lambda: f.read(4096), b''):
-                md5_hash.update(byte_block)
+        try:
+            with open(filename, 'rb') as f:
+                # Read and update hash in chunks of 4K
+                for byte_block in iter(lambda: f.read(4096), b''):
+                    md5_hash.update(byte_block)
+        except PermissionError as e:
+            raise SystemPermissionError(f'Failed to open DB: {filename}. {str(e)}')
 
         return md5_hash.hexdigest()
 
     def read_info_at_start(self) -> DBStartupAction:
+        """Read some metadata info at initialization
+
+        May raise:
+        - SystemPermissionError if there are permission errors when accessing the DB
+        """
         dbinfo = None
         action = DBStartupAction.NOTHING
         filepath = os.path.join(self.user_data_dir, DBINFO_FILENAME)
@@ -299,13 +320,22 @@ class DBHandler:
         self.conn.commit()
 
     def connect(self, password: str) -> None:
-        self.conn = sqlcipher.connect(  # pylint: disable=no-member
-            os.path.join(self.user_data_dir, 'rotkehlchen.db'),
-        )
+        """Connect to the DB using password
+
+        May raise:
+        - SystemPermissionError if we are unable to open the DB file,
+        probably due to permission errors
+        """
+        fullpath = os.path.join(self.user_data_dir, 'rotkehlchen.db')
+        try:
+            self.conn = sqlcipher.connect(fullpath)  # pylint: disable=no-member
+        except sqlcipher.OperationalError:  # pylint: disable=no-member
+            raise SystemPermissionError(
+                f'Could not open database file: {fullpath}. Permission errors?',
+            )
+
         self.conn.text_factory = str
         password_for_sqlcipher = _protect_password_sqlcipher(password)
-        # __import__("pdb").set_trace()
-
         self.conn.executescript(f'PRAGMA key="{password_for_sqlcipher}";')
         if self.sqlcipher_version == 3:
             script = f'PRAGMA key="{password_for_sqlcipher}"; PRAGMA kdf_iter={KDF_ITER};'
@@ -333,8 +363,9 @@ class DBHandler:
         return success, msg
 
     def disconnect(self) -> None:
-        self.conn.close()
-        self.conn = None
+        if hasattr(self, 'conn') and self.conn:
+            self.conn.close()
+            self.conn = None
 
     def export_unencrypted(self, temppath: FilePath) -> None:
         self.conn.executescript(
@@ -375,7 +406,12 @@ class DBHandler:
             self.conn.executescript(script)
             self.disconnect()
 
-        self.connect(password)
+        try:
+            self.connect(password)
+        except SystemPermissionError as e:
+            raise AssertionError(
+                f'Permission error when reopening the DB. {str(e)}. Should never happen here',
+            )
         self._run_actions_after_first_connection(password)
         # all went okay, remove the original temp backup
         os.remove(os.path.join(self.user_data_dir, 'rotkehlchen_temp_backup.db'))

--- a/rotkehlchen/errors.py
+++ b/rotkehlchen/errors.py
@@ -41,6 +41,10 @@ class RotkehlchenPermissionError(Exception):
     pass
 
 
+class SystemPermissionError(Exception):
+    pass
+
+
 class RemoteError(Exception):
     """Thrown when a remote API can't be reached or throws unexpected error"""
     pass

--- a/rotkehlchen/premium/sync.py
+++ b/rotkehlchen/premium/sync.py
@@ -113,7 +113,8 @@ class PremiumSyncManager():
 
         Returns true for success and False for error/failure
 
-        Can raise PremiumAuthenticationError due to an UnableToDecryptRemoteData
+        May raise:
+        - PremiumAuthenticationError due to an UnableToDecryptRemoteData
         coming from  decompress_and_decrypt_db. This happens when the given password
         does not match the one on the saved DB.
         """

--- a/rotkehlchen/rotkehlchen.py
+++ b/rotkehlchen/rotkehlchen.py
@@ -129,6 +129,7 @@ class Rotkehlchen():
         or can't authenticate with the server
         - DBUpgradeError if the rotki DB version is newer than the software or
         there is a DB upgrade and there is an error.
+        - SystemPermissionError if the directory or DB file can not be accessed
         """
         log.info(
             'Unlocking user',

--- a/rotkehlchen/rotkehlchen.py
+++ b/rotkehlchen/rotkehlchen.py
@@ -2,6 +2,7 @@
 
 import argparse
 import logging
+import os
 import time
 from typing import Any, Dict, List, Optional, Tuple, Union
 
@@ -19,7 +20,13 @@ from rotkehlchen.constants.assets import A_USD
 from rotkehlchen.data.importer import DataImporter
 from rotkehlchen.data_handler import DataHandler
 from rotkehlchen.db.settings import DBSettings, ModifiableDBSettings
-from rotkehlchen.errors import EthSyncError, InputError, PremiumAuthenticationError, RemoteError
+from rotkehlchen.errors import (
+    EthSyncError,
+    InputError,
+    PremiumAuthenticationError,
+    RemoteError,
+    SystemPermissionError,
+)
 from rotkehlchen.exchanges.manager import ExchangeManager
 from rotkehlchen.externalapis.alethio import Alethio
 from rotkehlchen.externalapis.cryptocompare import Cryptocompare
@@ -52,6 +59,12 @@ MAIN_LOOP_SECS_DELAY = 15
 
 class Rotkehlchen():
     def __init__(self, args: argparse.Namespace) -> None:
+        """Initialize the Rotkehlchen object
+
+        May Raise:
+        - SystemPermissionError if the given data directory's permissions
+        are not correct.
+        """
         self.lock = Semaphore()
         self.lock.acquire()
 
@@ -91,6 +104,10 @@ class Rotkehlchen():
 
         self.sleep_secs = args.sleep_secs
         self.data_dir = args.data_dir
+        if not os.access(self.data_dir, os.W_OK | os.R_OK):
+            raise SystemPermissionError(
+                f'The given data directory {self.data_dir} is not readable or writable',
+            )
         self.args = args
         self.msg_aggregator = MessagesAggregator()
         self.greenlet_manager = GreenletManager(msg_aggregator=self.msg_aggregator)

--- a/rotkehlchen/server.py
+++ b/rotkehlchen/server.py
@@ -15,6 +15,11 @@ log = RotkehlchenLogsAdapter(logger)
 
 class RotkehlchenServer():
     def __init__(self) -> None:
+        """Initializes the backend server
+        May raise:
+        - SystemPermissionError due to the given args containing a datadir
+        that does not have the correct permissions
+        """
         arg_parser = app_args(
             prog='rotki',
             description=(

--- a/rotkehlchen/tests/api/test_users.py
+++ b/rotkehlchen/tests/api/test_users.py
@@ -1,3 +1,4 @@
+import os
 from http import HTTPStatus
 from pathlib import Path
 from typing import Any, Dict
@@ -92,6 +93,25 @@ def test_user_creation(rotkehlchen_api_server, data_dir):
 
     # Check that the directory was created
     assert Path(data_dir / username / 'rotkehlchen.db').exists()
+
+
+@pytest.mark.parametrize('use_clean_caching_directory', [True])
+@pytest.mark.parametrize('start_with_logged_in_user', [False])
+def test_user_creation_permission_error(rotkehlchen_api_server, data_dir):
+    """Test that creating a user when data directory permissions are wrong is handled"""
+    os.chmod(data_dir, 0o200)
+    username = 'hania'
+    data = {
+        'name': username,
+        'password': '1234',
+    }
+    response = requests.put(api_url_for(rotkehlchen_api_server, "usersresource"), json=data)
+    assert_error_response(
+        response=response,
+        contained_in_msg='Failed to create directory for user: [Errno 13] Permission denied',
+        status_code=HTTPStatus.CONFLICT,
+    )
+    os.chmod(data_dir, 0o777)
 
 
 @pytest.mark.parametrize('start_with_logged_in_user', [False])
@@ -503,6 +523,52 @@ def test_user_set_premium_credentials(rotkehlchen_api_server, username):
     assert rotki.premium.credentials.serialize_secret() == VALID_PREMIUM_SECRET
     with patched_get:
         assert rotki.premium.is_active()
+
+
+@pytest.mark.parametrize('use_clean_caching_directory', [True])
+@pytest.mark.parametrize('start_with_logged_in_user', [False])
+def test_user_login_user_dir_permission_error(rotkehlchen_api_server, data_dir):
+    """Test that user login with userdir path permission errors is handled properly"""
+    username = 'a_user'
+    user_dir = Path(data_dir / username)
+    user_dir.mkdir()
+    db_path = Path(data_dir / username / 'rotkehlchen.db')
+    db_path.touch()
+    os.chmod(user_dir, 0o200)
+
+    data = {'action': 'login', "password": '123', 'sync_approval': 'unknown'}
+    response = requests.patch(
+        api_url_for(rotkehlchen_api_server, "usersbynameresource", name=username),
+        json=data,
+    )
+    assert_error_response(
+        response=response,
+        contained_in_msg=f'User {username} exists but DB is missing',
+        status_code=HTTPStatus.CONFLICT,
+    )
+
+
+@pytest.mark.parametrize('use_clean_caching_directory', [True])
+@pytest.mark.parametrize('start_with_logged_in_user', [False])
+def test_user_login_db_permission_error(rotkehlchen_api_server, data_dir):
+    """Test that user login with db path permission errors is handled properly"""
+    username = 'a_user'
+    user_dir = Path(data_dir / username)
+    user_dir.mkdir()
+    db_path = Path(data_dir / username / 'rotkehlchen.db')
+    db_path.touch()
+    os.chmod(db_path, 0o200)
+    data = {'action': 'login', "password": '123', 'sync_approval': 'unknown'}
+    response = requests.patch(
+        api_url_for(rotkehlchen_api_server, "usersbynameresource", name=username),
+        json=data,
+    )
+    assert_error_response(
+        response=response,
+        contained_in_msg='Could not open database file',
+        status_code=HTTPStatus.CONFLICT,
+    )
+    os.chmod(db_path, 0o777)
 
 
 def test_user_set_premium_credentials_errors(rotkehlchen_api_server, username):

--- a/rotkehlchen/tests/integration/test_premium.py
+++ b/rotkehlchen/tests/integration/test_premium.py
@@ -16,18 +16,10 @@ from rotkehlchen.tests.utils.premium import (
     VALID_PREMIUM_SECRET,
     assert_db_got_replaced,
     create_patched_requests_get_for_premium,
+    get_different_hash,
     setup_starting_environment,
 )
 from rotkehlchen.utils.misc import ts_now
-
-
-def get_different_hash(given_hash: str) -> str:
-    """Given the string hash get one that's different but has same length"""
-    new_hash = ''
-    for x in given_hash:
-        new_hash = new_hash + chr(ord(x) + 1)
-
-    return new_hash
 
 
 @pytest.mark.parametrize('start_with_valid_premium', [True])

--- a/rotkehlchen/tests/test_users.py
+++ b/rotkehlchen/tests/test_users.py
@@ -1,6 +1,11 @@
+import os
 import string
+from pathlib import Path
+
+import pytest
 
 from rotkehlchen.data_handler import DataHandler
+from rotkehlchen.errors import SystemPermissionError
 
 
 def _user_creation_and_login(username, password, data_dir, msg_aggregator):
@@ -52,3 +57,35 @@ def test_user_password_with_all_ascii(data_dir, function_scope_messages_aggregat
         data_dir=data_dir,
         msg_aggregator=function_scope_messages_aggregator,
     )
+
+
+@pytest.mark.parametrize('use_clean_caching_directory', [True])
+def test_users_query_permission_error(data_dir, function_scope_messages_aggregator):
+    not_allowed_dir = os.path.join(data_dir, 'notallowed')
+    allowed_user_dir = os.path.join(data_dir, 'allowed_user')
+    os.mkdir(not_allowed_dir)
+    os.chmod(not_allowed_dir, 0o200)
+    os.mkdir(allowed_user_dir)
+    Path(Path(allowed_user_dir) / 'rotkehlchen.db').touch()
+    handler = DataHandler(
+        data_directory=data_dir,
+        msg_aggregator=function_scope_messages_aggregator,
+    )
+    assert handler.get_users() == {'allowed_user': 'loggedout'}
+    # Change permissions back to that pytest cleanup can clean it
+    os.chmod(not_allowed_dir, 0o777)
+
+
+@pytest.mark.parametrize('use_clean_caching_directory', [True])
+def test_new_user_permission_error(data_dir, function_scope_messages_aggregator):
+    not_allowed_dir = os.path.join(data_dir, 'notallowed')
+    os.mkdir(not_allowed_dir)
+    os.chmod(not_allowed_dir, 0o200)
+    handler = DataHandler(
+        data_directory=not_allowed_dir,
+        msg_aggregator=function_scope_messages_aggregator,
+    )
+    with pytest.raises(SystemPermissionError):
+        handler.unlock(username='someuser', password='123', create_new=True)
+    # Change permissions back to that pytest cleanup can clean it
+    os.chmod(not_allowed_dir, 0o777)

--- a/rotkehlchen/tests/unit/test_app.py
+++ b/rotkehlchen/tests/unit/test_app.py
@@ -1,4 +1,10 @@
+import os
+
+import pytest
+
+from rotkehlchen.errors import SystemPermissionError
 from rotkehlchen.exchanges.manager import SUPPORTED_EXCHANGES
+from rotkehlchen.rotkehlchen import Rotkehlchen
 from rotkehlchen.tests.utils.factories import make_api_key, make_api_secret
 
 
@@ -38,3 +44,10 @@ def test_initializing_exchanges(uninitialized_rotkehlchen):
     )
 
     assert all(name in rotki.exchange_manager.connected_exchanges for name in SUPPORTED_EXCHANGES)
+
+
+def test_initializing_rotki_with_datadir_with_wrong_permissions(cli_args, data_dir):
+    os.chmod(data_dir, 0o200)
+    with pytest.raises(SystemPermissionError):
+        Rotkehlchen(args=cli_args)
+    os.chmod(data_dir, 0o777)

--- a/rotkehlchen/tests/unit/test_usage_analytics.py
+++ b/rotkehlchen/tests/unit/test_usage_analytics.py
@@ -1,3 +1,5 @@
+import sys
+
 from rotkehlchen.usage_analytics import create_usage_analytics
 from rotkehlchen.utils.misc import get_system_spec
 
@@ -9,5 +11,9 @@ def test_create_usage_analytics():
     assert 'system_release' in analytics
     assert 'system_version' in analytics
     assert analytics['rotki_version'] == get_system_spec()['rotkehlchen']
-    assert analytics['country'] != 'unknown'
-    assert analytics['city'] != 'unknown'
+    if sys.platform != 'darwin':
+        assert analytics['country'] != 'unknown'
+        assert analytics['city'] != 'unknown'
+    else:
+        assert analytics['country'] is not None
+        assert analytics['city'] is not None

--- a/rotkehlchen/tests/utils/premium.py
+++ b/rotkehlchen/tests/utils/premium.py
@@ -122,6 +122,15 @@ def create_patched_premium(
     return patched_premium_at_start, patched_premium_at_set, patched_get
 
 
+def get_different_hash(given_hash: str) -> str:
+    """Given the string hash get one that's different but has same length"""
+    new_hash = ''
+    for x in given_hash:
+        new_hash = new_hash + chr(ord(x) + 1)
+
+    return new_hash
+
+
 def setup_starting_environment(
         rotkehlchen_instance: Rotkehlchen,
         username: str,
@@ -150,7 +159,7 @@ def setup_starting_environment(
     if same_hash_with_remote:
         remote_hash = our_hash
     else:
-        remote_hash = 'a' + our_hash[1:]
+        remote_hash = get_different_hash(our_hash)
 
     if newer_remote_db:
         metadata_last_modify_ts = our_last_write_ts + 10

